pinc/maybe_mail.inc

```diff
@@ -4,6 +4,135 @@ include_once($relPath.'misc.inc'); // html_safe()
 include_once($relPath.'Project.inc');
 include_once($relPath.'User.inc');
 
+use PHPMailer\PHPMailer\PHPMailer;
+use PHPMailer\PHPMailer\SMTP;
+use PHPMailer\PHPMailer\Exception;
+
+function maybe_email($to, $subject, $message, $from = null, $reply_to = null)
+// If this is for real, send the message.
+// If we're testing, just report what would have been sent.
+// Optional $from and $reply_to arguments may either take the form
+// "user@example.com" or "Example User <user@example.com>"
+{
+    global $testing, $auto_email_addr, $site_name, $site_url;
+
+    $from = $from ?? $auto_email_addr;
+    $reply_to = $reply_to ?? $auto_email_addr;
+
+    $success = true;
+
+    // Append a standard footer to all emails sent out by the system.
+    if (!endswith($message, "\n")) {
+        $message .= "\n";
+    }
+
+    $message .= implode("\n", [
+        "",
+        "--  ",
+        "$site_name  ",
+        $site_url,
+        "",
+        // We intentionally don't translate this message because we don't know
+        // what the recipient's language is at this point.
+        "This is an automated message.",
+    ]);
+
+    $Parsedown = new ParsedownExtra();
+    $Parsedown->setSafeMode(true);
+    $html_message = $Parsedown->text($message);
+
+    // $mail_mode variable controls whether and how emails are sent
+    // =0 - don't send mails (usual testing mode, unless testing emails)
+    // >0 - send mails normally (usual production mode)
+    // -1 - use localhost:2025 (for testing emails)
+    //      Example test setup: Use https://github.com/tomwardill/FakeEmail
+    //          pip install fakeemail
+    //          fakeemail 2025 8080 0.0.0.0
+    //          Open http://mydomain.org:8080/ in browser to monitor emails
+    $mail_mode = 0;
+
+    if ($testing) {
+        echo "\n<hr>\n";
+        echo "\$testing is $testing. If it were FALSE, the following Markdown mail would have been sent:\n";
+        echo "<pre style='white-space: pre-wrap'>\n";
+        echo html_safe("To: $to\n");
+        echo html_safe("Subject: $subject\n");
+        echo html_safe("From: $from\n");
+        echo html_safe("Reply-To: $reply_to\n");
+        echo "</pre>";
+        echo "<pre style='white-space: pre-wrap'>";
+        echo html_safe("$message\n");
+        echo "</pre>\n";
+        echo "<hr>\n";
+        echo "$html_message\n"   ;
+        echo "<hr>\n";
+        // Uncomment below to test emails using localhost:2025
+        $mail_mode = -1;
+    } else {
+        $mail_mode = 1;
+    }
+
+    if ($mail_mode === 0) {
+        return success;    // Nothing left to do!
+    }
+
+    $mail = new PHPMailer(true);
+    try {
+        // Server settings
+        $mail->isSMTP();
```

Comment:
Is it possible to have PHPMailer use the PHP `mail()` function to send email? If it is, that might be a good first version as we know that already works at pgdp.net and we don't have to wire through additional email settings.

Reply:
I don't know - I didn't see one straight away, as it occurred to me when I was struggling to get it working in the first place (because I had security enabled which FakeEmail didn't want). I'll have another look.

Reply:
Doesn't look like it. Let me give some thought on how we can pass in config info because that's going to be a bit of a challenge I think.

Reply:
I have a commit (pushed up to my repo): https://github.com/cpeel/dproofreaders/commit/224eb4cccd900009a6098296c3b417e0130d5b76 that adds dynamic configuration. I have updated `~cpeel/bin/TEST_setup.sh` and `~cpeel/bin/BRANCH_setup.sh` to use the FakeEmail port, so you should be able to update your `site_vars.php` to use these new settings:
```bash
git checkout email
./SETUP/configure ~cpeel/bin/TEST_setup.sh .
```

You're welcome to take or drop whatever parts of this you want.